image.go

```diff
@@ -393,15 +393,23 @@ func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.System
 			rc.Close()
 			return nil, errors.Wrapf(err, "error opening file for %s", what)
 		}
-		destHasher := digest.Canonical.Digester()
+
 		counter := ioutils.NewWriteCounter(layerFile)
-		multiWriter := io.MultiWriter(counter, destHasher.Hash())
-		// Compress the layer, if we're recompressing it.
-		writeCloser, err := archive.CompressStream(multiWriter, i.compression)
-		if err != nil {
-			layerFile.Close()
-			rc.Close()
-			return nil, errors.Wrapf(err, "error compressing %s", what)
+		var destHasher digest.Digester
+		var writeCloser io.WriteCloser
+		if i.compression != archive.Uncompressed {
+			destHasher = digest.Canonical.Digester()
+			multiWriter := io.MultiWriter(counter, destHasher.Hash())
+			// Compress the layer, if we're recompressing it.
+			writeCloser, err = archive.CompressStream(multiWriter, i.compression)
+			if err != nil {
+				layerFile.Close()
+				rc.Close()
+				return nil, errors.Wrapf(err, "error compressing %s", what)
+			}
+		} else {
+			destHasher = srcHasher
+			writeCloser = ioutils.NopWriteCloser(counter)
```

Comment:
`archive.CompressStream(…, archive.Uncompressed)` actually wraps the input in a buffered writer. Does that matter for performance either way?

( The original code in https://github.com/moby/moby/pull/7542 doesn’t really talk about this case; it might well not be necessary to use a buffer. )

Reply:
I left it out to avoid an additional writer in the chain but at least on my machine I dont think it makes much difference with or without it. If we prefer we can keep it in the chain.

Reply:
Reading the linked PR its probably better to preserve it as it was added for network and other paths that I did not test, and it doesnt slow down the fast disk case I was testing.

Reply:
I think the discussed network path is not involved here - we read from c/storage and write to a local file (vs. streaming read from storage -> compress -> push over HTTP to registry). So I don’t think that old conversation is all that informative.

Anyway, sure, the updated version of this PR makes this question moot.